File: prospr/exact/depth_first.py
#!/usr/bin/env python3
"""
File:       depth_first.py
Author:     Okke van Eck

Description:    This file defines the depth-first search algorithm for folding a
                Protein object.
"""

import numpy as np


def depth_first(protein):
    """
    Depth-first algorithm for folding a protein.
    :param Protein  protein:    Protein object to process.
    :return int:                Best found score.
    """
    max_length = len(protein.get_sequence())

    # First two aminos are placed to prevent symmetry. These are not placed on
    # the stack since it may not be backtracked.
    if max_length >= 1:
        protein.place_amino(0)
    if max_length >= 2:
        protein.place_amino(2)
    if max_length < 3:
        return 0

    # Setup the stack as if the protein is one long upwards going chain. The
    # remaining moves are all the positive directions of all dimensions to
    # prevent symmetry over all axis.
    dim = protein.get_dim()
    stack_size = max_length - 2
    move = 2

    stack = np.zeros((stack_size, 2 * dim - 1), dtype=np.int64)
    stack_pointer = stack_size - 1

    for i in range(max_length - 2):
        protein.place_amino(move)
        idx = 0

        for j in range(-dim, dim):
            if j != 0 and j != move:
                stack[i][idx] = j
                idx += 1

    # Init variables for run.
    moves = np.array([m for m in range(-dim, dim) if m != 0], dtype=np.int64)
    best_score = 1
    best_hash = None

    while stack:
        # Check score if bottom of tree is reached and save best protein.
        # Otherwise continue with next amino and possible moves.
        if protein.get_cur_len() == max_length:
            score = protein.get_score()

            if score < best_score:
                best_score = score
                best_hash = protein.hash_fold()

            remaining_moves = np.zeros((2 * dim - 1), dtype=np.int64)
        else:
            remaining_moves = moves[moves != -move]
            remaining_moves.reshape((2 * dim - 1))

        # Loop till an amino has been placed or no new combinations are left.
        placed_amino = False

        while not placed_amino:
            # Try all remaining moves for the current amino and place if valid.
            while not placed_amino and remaining_moves:
                # Pop move from remaining moves.
                move = remaining_moves[0]
                remaining_moves = remaining_moves[remaining_moves != move]
                remaining_moves.reshape((2 * dim - 1))

                # Place amino, update length and save the move on the stack
                # with the remaining possible moves, if valid.
                if protein.is_valid(move):
                    protein.place_amino(move)
                    stack[stack_pointer] = remaining_moves
                    stack_pointer += 1
                    placed_amino = True

            # Backtrack if the current amino has no possible moves left.
            # while not placed_amino and not remaining_moves and stack:
            while placed_amino == False and remaining_moves.size == 0 and stack:
                remaining_moves = stack[-1]
                stack_pointer -= 1
                protein.remove_amino()

            # Check if all possible combinations have been checked.
            if not remaining_moves:
                break

    protein.set_hash(best_hash)

    return best_score, best_hash

File: prospr/core/exact/depth_first/depth_first.cpp
/* File:       depth_fisrt.cpp
 * Author:     Okke van Eck
 *
 * Description:    Source file for a depth_first search function.
 */

#include "depth_first.h"
#include <stack>
#include <vector>
#include <algorithm>

#include <iostream>


/* Initialize the stack and all_moves vector. */
void initialize_vars(Protein* protein,
                     std::stack<std::vector<int>>* dfs_stack,
                     std::vector<int>* all_moves,
                     int max_length,
                     int first_move) {
    int dim = protein->get_dim();

    /*  Setup all_moves for stack initialization. */
    for (int i = 1; i <= dim; i++) {
        if (i == first_move)
            continue;

        all_moves->push_back(i);
    }

    /* Setup stack, limiting negative movement to prevent symmetry. */
    for (int i = 0; i < max_length - 2; i++) {
        protein->place_amino(2);
        dfs_stack->push(*all_moves);
    }

    /* Setup all_moves for further usage. */
    all_moves->clear();

    for (int i = -dim; i <= dim; i++) {
        if (i == 0)
            continue;

        all_moves->push_back(i);
    }
}

/* A depth-first search function for finding a minimum energy conformation. */
Protein depth_first(Protein protein) {
    int max_length = protein.get_sequence().length();

    /* The first two amino acids are fixed in order to prevent symmetry. */
    if (max_length >= 1)
        protein.place_amino(0);
    if (max_length >= 2)
        protein.place_amino(2);
    if (max_length < 3)
        return protein;

    std::cout << "First ifs worked!\n";

    /* Setup all_moves for initializing the stack. */
    int move = 2;
    std::vector<int> all_moves;
    std::stack<std::vector<int>> dfs_stack;
    initialize_vars(&protein, &dfs_stack, &all_moves, max_length, move);

    std::cout << "Setup initials worked!\n";

//    std::cout << "\ndfs_stack:\n";
//    while(!dfs_stack.empty()) {
//        std::vector<int> v = dfs_stack.top();
//
//        for (auto i = 0; i < v.size(); i++) {
//            std::cout << v.at(i) << ' ';
//        }
//
//        std::cout << "\n";
//        dfs_stack.pop();
//    }
//
//    std::cout << "\nall_moves:\n";
//    for (auto i = 0; i < all_moves.size(); i++) {
//        std::cout << all_moves.at(i) << ' ';
//    }
//    std::cout << "\n";
//
//    std::cout << "\nmax_length:" << max_length << "\n";
//    std::cout << "\nmove:" << move << "\n";


    /* Declare and set variables for the depth-first search. */
    int best_score = 1;
    std::vector<int> best_hash;
    std::vector<int> remaining_moves;
    int score;
    bool placed_amino;

    while (!dfs_stack.empty()) {
        /* Check if the bottom of the tree is reached and store best folds.
         * Continue with next amino acid otherwise.
         */
         std::cout << protein.get_cur_len() << "\n";
        if (protein.get_cur_len() == max_length) {
            score = protein.get_score();

            if (score < best_score) {
                best_score = score;
                best_hash = protein.hash_fold();
            }

            remaining_moves.clear();
        } else {
            remaining_moves = all_moves;
            remaining_moves.erase(std::remove(remaining_moves.begin(),
                                              remaining_moves.end(), -move),
                                  remaining_moves.end());
        }

        /* Loop till an amino acid is placed or no new combinations are left. */
        placed_amino = false;

        while (!placed_amino) {
            /* Try all remaining moves for the current amino acid. */
            while (!placed_amino && !remaining_moves.empty()) {
                move = remaining_moves.back();
                remaining_moves.pop_back();

                /* Place amino acid if valid and update stack accordingly. */
                if (protein.is_valid(move)) {
                    protein.place_amino(move);
                    dfs_stack.push(remaining_moves);
                    placed_amino = true;
                }
            }

            /* Backtrack if the current amino acid has no moves left. */
            while (!placed_amino && remaining_moves.empty() &&
                   !dfs_stack.empty()) {
                remaining_moves = dfs_stack.top();
                dfs_stack.pop();
                protein.remove_amino();
            }

            /* Check if there are no moves left. */
            if (remaining_moves.empty())
                break;
        }
    }

    protein.set_hash(best_hash);

    return protein;
}
